### PR TITLE
[release/8.0-staging] Call the Copy Constructor for stack arguments in C++/CLI on x86

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
This is a port of https://github.com/dotnet/runtime/pull/100050 to .NET 8.

/cc @jeffschwMSFT @agocke @jkotas 

## Customer Impact

- [x] Customer reported
- [ ] Found internally

See user reported issue https://github.com/dotnet/runtime/issues/100033.

> The DTS bug contains the repro. Essentially, a copy ctor of std::vector<>::iterator is not called when pushing the argument onto the stack for the call to the native method, which takes an argument of type iterator. This leads to STL's iterator debugging book-keeping code to later assert.
>
> This happens only for x86 and does not repro in the desktop framework where the copy ctor is correctly called.

The workaround is to disable debug iterators or change the legacy code to avoid the problematic pattern. Both of these are unacceptable long-term in-production fixes.

## Regression

- [x] Yes
- [ ] No

Regression introduced from .NET Framework. Support was added in .NET Core 3.1 and included the regression - https://github.com/dotnet/coreclr/pull/22805.

## Testing

A test was added that validates the issue in a C++/CLI scenario.

## Risk

Low

### CODE REVIEW

File: src/coreclr/vm/ilmarshalers.cpp [hunk 1]
```diff
@@ -3394,6 +3394,7 @@ ILCriticalHandleMarshaler::ReturnOverride(
     return OVERRIDDEN;
 } // ILCriticalHandleMarshaler::ReturnOverride
 
+#if defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:3397]:
```suggestion
```

File: src/coreclr/vm/ilmarshalers.cpp [hunk 2]
```diff
@@ -3459,6 +3460,36 @@ MarshalerOverrideStatus ILBlittableValueClassWithCopyCtorMarshaler::ArgumentOver
 #ifdef TARGET_X86
         pslIL->SetStubTargetArgType(&locDesc);              // native type is the value type
         pslILDispatch->EmitLDLOC(dwNewValueTypeLocal);      // we load the local directly
+
+        // Record this argument's stack slot in the copy constructor chain so we can correctly invoke the copy constructor.
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:3464]:
```suggestion
#if defined(TARGET_WINDOWS)
        // Record this argument's stack slot in the copy constructor chain so we can correctly invoke the copy constructor.
```

File: src/coreclr/vm/ilmarshalers.cpp [hunk 3]
```diff
@@ -3459,6 +3460,36 @@ MarshalerOverrideStatus ILBlittableValueClassWithCopyCtorMarshaler::ArgumentOver
...
+        if (pargs->mm.m_pDtor)
+        {
+            pslIL->EmitLDLOCA(ctorCookie);
+            pslIL->EmitLDFTN(pslIL->GetToken(pargs->mm.m_pDtor));
+            pslIL->EmitSTFLD(pslIL->GetToken(CoreLibBinder::GetField(FIELD__COPY_CONSTRUCTOR_COOKIE__DESTRUCTOR)));
+        }
+
+        pslIL->EmitLDLOCA(psl->GetCopyCtorChainLocalNum());
+        pslIL->EmitLDLOCA(ctorCookie);
+        pslIL->EmitCALL(METHOD__COPY_CONSTRUCTOR_CHAIN__ADD, 2, 0);
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:3491]:
```suggestion
        pslIL->EmitCALL(METHOD__COPY_CONSTRUCTOR_CHAIN__ADD, 2, 0);
#endif // TARGET_WINDOWS
```

File: src/coreclr/vm/ilmarshalers.cpp [hunk 4]
```diff
@@ -3477,9 +3508,7 @@ MarshalerOverrideStatus ILBlittableValueClassWithCopyCtorMarshaler::ArgumentOver
 
         DWORD       dwNewValueTypeLocal;
         dwNewValueTypeLocal = pslIL->NewLocal(locDesc);
-        pslILDispatch->EmitLDARG(argidx);
-        pslILDispatch->EmitSTLOC(dwNewValueTypeLocal);
-        pslILDispatch->EmitLDLOCA(dwNewValueTypeLocal);
+        pslILDispatch->EmitLDARGA(argidx);
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:3511]:
```suggestion
#ifdef TARGET_WINDOWS
        pslILDispatch->EmitLDARGA(argidx);
#else
        pslILDispatch->EmitLDARG(argidx);
        pslILDispatch->EmitSTLOC(dwNewValueTypeLocal);
        pslILDispatch->EmitLDLOCA(dwNewValueTypeLocal);
#endif
```

File: src/coreclr/vm/ilmarshalers.cpp [hunk 5]
```diff
@@ -3491,6 +3520,7 @@ MarshalerOverrideStatus ILBlittableValueClassWithCopyCtorMarshaler::ArgumentOver
         return OVERRIDDEN;
     }
 }
+#endif // defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.cpp:3523]:
```suggestion
```

File: src/coreclr/vm/ilmarshalers.h [hunk 6]
```diff
@@ -2889,6 +2889,7 @@ class ILBlittableLayoutClassMarshaler : public ILMarshaler
     void EmitConvertContentsNativeToCLR(ILCodeStream* pslILEmit) override;
 };
 
+#if defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.h:2892]:
```suggestion
```

File: src/coreclr/vm/ilmarshalers.h [hunk 7]
```diff
@@ -2922,6 +2923,7 @@ class ILBlittableValueClassWithCopyCtorMarshaler : public ILMarshaler
 
 
 };
+#endif // defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/ilmarshalers.h:2926]:
```suggestion
```

File: src/coreclr/vm/mlinfo.cpp [hunk 8]
```diff
@@ -2371,6 +2371,7 @@ MarshalInfo::MarshalInfo(Module* pModule,
                     {
                         if (fNeedsCopyCtor && !IsFieldScenario()) // We don't support automatically discovering copy constructors for fields.
                         {
+#if defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mlinfo.cpp:2374]:
```suggestion
```

File: src/coreclr/vm/mlinfo.cpp [hunk 9]
```diff
@@ -2380,6 +2381,10 @@ MarshalInfo::MarshalInfo(Module* pModule,
                             m_args.mm.m_pCopyCtor = pCopyCtor;
                             m_args.mm.m_pDtor = pDtor;
                             m_type = MARSHAL_TYPE_BLITTABLEVALUECLASSWITHCOPYCTOR;
+#else // !defined(TARGET_WINDOWS)
+                            m_resID = IDS_EE_BADMARSHAL_BADMANAGED;
+                            IfFailGoto(E_FAIL, lFail);
+#endif // defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mlinfo.cpp:2387]:
```suggestion
```

File: src/coreclr/vm/mlinfo.cpp [hunk 10]
```diff
@@ -3109,7 +3114,9 @@ bool MarshalInfo::IsValueClass(MarshalType mtype)
     {
     case MARSHAL_TYPE_BLITTABLEVALUECLASS:
     case MARSHAL_TYPE_VALUECLASS:
+#if defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mlinfo.cpp:3117]:
```suggestion
```

File: src/coreclr/vm/mlinfo.cpp [hunk 11]
```diff
@@ -3109,7 +3114,9 @@ bool MarshalInfo::IsValueClass(MarshalType mtype)
...
     case MARSHAL_TYPE_BLITTABLEVALUECLASSWITHCOPYCTOR:
+#endif // defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mlinfo.cpp:3119]:
```suggestion
```

File: src/coreclr/vm/mlinfo.cpp [hunk 12]
```diff
@@ -3593,7 +3600,9 @@ DispParamMarshaler *MarshalInfo::GenerateDispParamMarshaler()
         case MARSHAL_TYPE_BLITTABLEVALUECLASS:
         case MARSHAL_TYPE_BLITTABLEPTR:
         case MARSHAL_TYPE_LAYOUTCLASSPTR:
+#if defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mlinfo.cpp:3603]:
```suggestion
```

File: src/coreclr/vm/mlinfo.cpp [hunk 13]
```diff
@@ -3593,7 +3600,9 @@ DispParamMarshaler *MarshalInfo::GenerateDispParamMarshaler()
...
         case MARSHAL_TYPE_BLITTABLEVALUECLASSWITHCOPYCTOR:
+#endif // defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mlinfo.cpp:3605]:
```suggestion
```

File: src/coreclr/vm/mtypes.h [hunk 14]
```diff
@@ -79,7 +79,10 @@ DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_VALUECLASS,                      ValueClassMa
 
 DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_REFERENCECUSTOMMARSHALER,        ReferenceCustomMarshaler)
 DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_ARGITERATOR,                     ArgIteratorMarshaler)
+
+#if defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mtypes.h:83]:
```suggestion
```

File: src/coreclr/vm/mtypes.h [hunk 15]
```diff
@@ -79,7 +79,10 @@ DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_VALUECLASS,                      ValueClassMa
...
 DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_BLITTABLEVALUECLASSWITHCOPYCTOR, BlittableValueClassWithCopyCtorMarshaler)
+#endif // defined(TARGET_WINDOWS)
```
Comment (jkotas) [inline src/coreclr/vm/mtypes.h:85]:
```suggestion
```

**Review by jkotas — APPROVED** [review]
I would avoid the additional cleanup in the servicing backport.

**Comment by AaronRobinsonMSFT** [comment]
> I would avoid the additional cleanup in the servicing backport.

Sounds good. I wasn't sure how much effort we wanted to narrow the changes. I just tried applying them without a local build. If there is any fallout, I will look again tomorrow.

**Review by jeffschwMSFT — APPROVED** [review]
approved. we will take for consideration in 8.0.x

**Review by agocke — APPROVED** [review]

**Pull request merged**